<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Pbase example data</title>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: #990073
   }

   pre .number {
     color: #099;
   }

   pre .comment {
     color: #998;
     font-style: italic
   }

   pre .keyword {
     color: #900;
     font-weight: bold
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: #d14;
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>

<!-- MathJax scripts -->
<script type="text/javascript" src="https://c328740.ssl.cf1.rackcdn.com/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>


<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}

pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<!--
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{Pbase-data}
-->

<!-- --- -->

<!-- title: "Pbase example data" -->

<!-- author: "Laurent Gatto - <lg390@cam.ac.uk>&ldquo; &ndash;>

<!-- output: -->

<!--   html_document: -->

<!--     toc: true -->

<!--     theme: united -->

<!-- --- -->

<h1>Pbase example data</h1>

<p>Laurent Gatto - <a href="mailto:lg390@cam.ac.uk">lg390@cam.ac.uk</a></p>

<p>The original data is a 10 fmol
<a href="http://www.piercenet.com/product/peptide-retention-time-calibration-mixture">Peptide Retention Time Calibration Mixture</a>
spiked into 50 ng HeLa background acquired on a Thermo Orbitrap Q
Exactive instrument. A restricted set of high scoring human proteins
from the UniProt release 2014_06 were searched using the MSGF+ engine.</p>

<h2>The fasta database</h2>

<pre><code class="r">library(&quot;Biostrings&quot;)
</code></pre>

<pre><code>## Loading required package: XVector
</code></pre>

<pre><code class="r">fafile &lt;- system.file(&quot;extdata/HUMAN_2014_06-09prots.fasta&quot;, package = &quot;Pbase&quot;)
fa &lt;- readAAStringSet(fafile)
fa
</code></pre>

<pre><code>##   A AAStringSet instance of length 9
##     width seq                                          names               
## [1]   375 MDDDIAALVVDNGSGMCKAGF...WISKQEYDESGPSIVHRKCF sp|P60709|ACTB_HU...
## [2]   664 METPSQRRATRSGAQASSTPL...SYLLGNSSPRTQSPQNCSIM sp|P02545|LMNA_HU...
## [3]   417 MSLSNKLTLDKLDVKGKRVVM...ASLELLEGKVLPGVDALSNI sp|P00558|PGK1_HU...
## [4]  2602 MPVTEKDLAEDAPWKKIQQNT...LAVKWGEEHIPGSPFHVTVP sp|O75369|FLNB_HU...
## [5]   911 MVDYHAANQSYQYGPSSAGNG...VPGALDYKSFSTALYGESDL sp|O43707|ACTN4_H...
## [6]  2624 MFRRARLSVKPNVRPGVGARG...ATTVSEYFFNDIFIEVDETE sp|A6H8Y1|BDP1_HU...
## [7]  3374 MSPESGHSRIFEATAGPNKPE...TLSKDSLSNGVPSGRQAEFS sp|A4UGR9|XIRP2_H...
## [8]   364 MPYQYPALTPEQKKELSDIAH...PSGQAGAAASESLFVSNHAY sp|P04075|ALDOA_H...
## [9]   418 MARRKPEGSSFNMTHLSMAMA...PSGQAGAAASESLFVSNHAY sp|P04075-2|ALDOA...
</code></pre>

<h2>The PSM data</h2>

<pre><code class="r">library(&quot;mzID&quot;)
idfile &lt;- system.file(&quot;extdata/Thermo_Hela_PRTC_1_selected.mzid&quot;, package = &quot;Pbase&quot;)
id &lt;- flatten(mzID(idfile))
</code></pre>

<pre><code>## reading Thermo_Hela_PRTC_1_selected.mzid... DONE!
</code></pre>

<pre><code class="r">dim(id)
</code></pre>

<pre><code>## [1] 137  39
</code></pre>

<pre><code class="r">head(id)
</code></pre>

<pre><code>##   spectrumid acquisitionnum passthreshold rank calculatedmasstocharge
## 1  index=173            173          TRUE    1      1136.574462890625
## 2  index=173            173          TRUE    1      1136.574462890625
## 3  index=163            163          TRUE    1      1136.574462890625
## 4  index=163            163          TRUE    1      1136.574462890625
## 5    index=8              8          TRUE    1      469.2348937988281
## 6    index=8              8          TRUE    1      469.2348937988281
##   experimentalmasstocharge chargestate ms-gf:denovoscore ms-gf:evalue
## 1        1137.066650390625           2               132    2.597e-18
## 2        1137.066650390625           2               132    2.597e-18
## 3        1136.574462890625           2               230    4.943e-17
## 4        1136.574462890625           2               230    4.943e-17
## 5       469.23480224609375           2                49    3.824e-06
## 6       469.23480224609375           2                49    3.824e-06
##   ms-gf:rawscore ms-gf:specevalue assumeddissociationmethod
## 1            118        2.277e-22                       CID
## 2            118        2.277e-22                       CID
## 3            186        4.333e-21                       CID
## 4            186        4.333e-21                       CID
## 5             44        3.374e-10                       CID
## 6             44        3.374e-10                       CID
##   ctermioncurrentratio explainedioncurrentratio isotopeerror meanerrorall
## 1            0.5061673                0.5888804            1    17.757027
## 2            0.5061673                0.5888804            1    17.757027
## 3           0.48202217               0.57532895            0    18.029726
## 4           0.48202217               0.57532895            0    18.029726
## 5             0.285157               0.44413796            0    4.1399555
## 6             0.285157               0.44413796            0    4.1399555
##   meanerrortop7 meanrelerrorall meanrelerrortop7 ms2ioncurrent
## 1     4.3787823     -15.7646885        -3.255431      526864.0
## 2     4.3787823     -15.7646885        -3.255431      526864.0
## 3     5.0083427       -16.59046        -3.333969     3342169.2
## 4     5.0083427       -16.59046        -3.333969     3342169.2
## 5      4.087293      -2.5620477       -2.0585544     8156462.0
## 6      4.087293      -2.5620477       -2.0585544     8156462.0
##   ntermioncurrentratio nummatchedmainions stdeverrorall stdeverrortop7
## 1           0.08271315                 25     23.135044       3.582861
## 2           0.08271315                 25     23.135044       3.582861
## 3           0.09330681                 30     28.647552      2.9929223
## 4           0.09330681                 30     28.647552      2.9929223
## 5           0.15898097                  9     2.7658114      2.7357001
## 6           0.15898097                  9     2.7658114      2.7357001
##   stdevrelerrorall stdevrelerrortop7                  pepseq modified
## 1         24.53603          4.627396 GVVPLAGTNGETTTQGLDGLSER    FALSE
## 2         24.53603          4.627396 GVVPLAGTNGETTTQGLDGLSER    FALSE
## 3        29.504406          4.788082 GVVPLAGTNGETTTQGLDGLSER    FALSE
## 4        29.504406          4.788082 GVVPLAGTNGETTTQGLDGLSER    FALSE
## 5         4.269058          4.466808                AAQEEYVK    FALSE
## 6         4.269058          4.466808                AAQEEYVK    FALSE
##   modification isdecoy post pre end start               accession length
## 1         &lt;NA&gt;   FALSE    C   K 134   112   sp|P04075|ALDOA_HUMAN    364
## 2         &lt;NA&gt;   FALSE    C   K 188   166 sp|P04075-2|ALDOA_HUMAN    418
## 3         &lt;NA&gt;   FALSE    C   K 134   112   sp|P04075|ALDOA_HUMAN    364
## 4         &lt;NA&gt;   FALSE    C   K 188   166 sp|P04075-2|ALDOA_HUMAN    418
## 5         &lt;NA&gt;   FALSE    R   K 330   323   sp|P04075|ALDOA_HUMAN    364
## 6         &lt;NA&gt;   FALSE    R   K 384   377 sp|P04075-2|ALDOA_HUMAN    418
##                                                              description
## 1    Fructose-bisphosphate aldolase A OS=Homo sapiens GN=ALDOA PE=1 SV=2
## 2 Isoform 2 of Fructose-bisphosphate aldolase A OS=Homo sapiens GN=ALDOA
## 3    Fructose-bisphosphate aldolase A OS=Homo sapiens GN=ALDOA PE=1 SV=2
## 4 Isoform 2 of Fructose-bisphosphate aldolase A OS=Homo sapiens GN=ALDOA
## 5    Fructose-bisphosphate aldolase A OS=Homo sapiens GN=ALDOA PE=1 SV=2
## 6 Isoform 2 of Fructose-bisphosphate aldolase A OS=Homo sapiens GN=ALDOA
##                    spectrumFile                databaseFile
## 1 Thermo_Hela_PRTC_selected.mgf HUMAN_2014_06-09prots.fasta
## 2 Thermo_Hela_PRTC_selected.mgf HUMAN_2014_06-09prots.fasta
## 3 Thermo_Hela_PRTC_selected.mgf HUMAN_2014_06-09prots.fasta
## 4 Thermo_Hela_PRTC_selected.mgf HUMAN_2014_06-09prots.fasta
## 5 Thermo_Hela_PRTC_selected.mgf HUMAN_2014_06-09prots.fasta
## 6 Thermo_Hela_PRTC_selected.mgf HUMAN_2014_06-09prots.fasta
</code></pre>

<h2>The Proteins object</h2>

<pre><code class="r">library(&quot;Pbase&quot;)
p &lt;- Proteins(fafile)
p &lt;- addIdentificationData(p, idfile)
</code></pre>

<pre><code>## reading Thermo_Hela_PRTC_1_selected.mzid... DONE!
</code></pre>

<pre><code class="r">p
</code></pre>

<pre><code>## S4 class type     : Proteins
## Class version     : 0.1
## Created           : Wed Jul 16 02:45:51 2014
## Number of Proteins: 9
## Sequences:
##   [1] A4UGR9 [2] A6H8Y1 ... [8] P04075-2 [9] P60709
## Sequence features:
##   [1] DB [2] AccessionNumber ... [10] Comment [11] Filename
## Peptide features:
##   [1] DB [2] AccessionNumber ... [46] spectrumFile [47] databaseFile
</code></pre>

<p>A <code>Proteins</code> object is composed of a set of protein sequences
accessible with the <code>aa</code> accessor as well as an optional set of
peptides features that are mapped as coordinates along the proteins,
available with <code>pranges</code>. The actual peptide sequences can be extraced
with <code>pfeatures</code>.</p>

<pre><code class="r">aa(p)
</code></pre>

<pre><code>##   A AAStringSet instance of length 9
##     width seq                                          names               
## [1]  3374 MSPESGHSRIFEATAGPNKPE...TLSKDSLSNGVPSGRQAEFS A4UGR9
## [2]  2624 MFRRARLSVKPNVRPGVGARG...ATTVSEYFFNDIFIEVDETE A6H8Y1
## [3]   911 MVDYHAANQSYQYGPSSAGNG...VPGALDYKSFSTALYGESDL O43707
## [4]  2602 MPVTEKDLAEDAPWKKIQQNT...LAVKWGEEHIPGSPFHVTVP O75369
## [5]   417 MSLSNKLTLDKLDVKGKRVVM...ASLELLEGKVLPGVDALSNI P00558
## [6]   664 METPSQRRATRSGAQASSTPL...SYLLGNSSPRTQSPQNCSIM P02545
## [7]   364 MPYQYPALTPEQKKELSDIAH...PSGQAGAAASESLFVSNHAY P04075
## [8]   418 MARRKPEGSSFNMTHLSMAMA...PSGQAGAAASESLFVSNHAY P04075-2
## [9]   375 MDDDIAALVVDNGSGMCKAGF...WISKQEYDESGPSIVHRKCF P60709
</code></pre>

<pre><code class="r">pranges(p)
</code></pre>

<pre><code>## IRangesList of length 9
## $A4UGR9
## IRanges of length 36
##      start  end width  names
## [1]   2743 2760    18 A4UGR9
## [2]    307  318    12 A4UGR9
## [3]   1858 1870    13 A4UGR9
## [4]   1858 1870    13 A4UGR9
## [5]   1699 1708    10 A4UGR9
## ...    ...  ...   ...    ...
## [32]  2082 2094    13 A4UGR9
## [33]    20   31    12 A4UGR9
## [34]  1712 1729    18 A4UGR9
## [35]    48   61    14 A4UGR9
## [36]  2743 2756    14 A4UGR9
## 
## ...
## &lt;8 more elements&gt;
</code></pre>

<pre><code class="r">pfeatures(p)
</code></pre>

<pre><code>## AAStringSetList of length 9
## [[&quot;A4UGR9&quot;]] QEITQNKSFFSSVKESQR LPVPKDVYSKQR ... QEITQNKSFFSSVK
## [[&quot;A6H8Y1&quot;]] EDAEQVALEVDLNQKKRR ... ARLSVKPNVRPGVGARGSTASNPQRGR
## [[&quot;O43707&quot;]] QQRKTFTAWCNSHLR CQKICDQWDALGSLTHSR ... VGWEQLLTTIAR
## [[&quot;O75369&quot;]] DLDIIDNYDYSHTVK PFDLVIPFAVRK ... VQAQGPGLKEAFTNK
## [[&quot;P00558&quot;]] ELNYFAKALESPER DLMSKAEK QIVWNGPVGVFEWEAFAR FHVEEEGKGKDASGNK GTKALMDEVVK
## [[&quot;P02545&quot;]] METPSQRRATR DTSRRLLAEKEREMAEMR ... RATRSGAQASSTPLSPTR
## [[&quot;P04075&quot;]] GVVPLAGTNGETTTQGLDGLSER ... TVPPAVTGITFLSGGQSEEEASINLNAINK
## [[&quot;P04075-2&quot;]] GVVPLAGTNGETTTQGLDGLSER ... TVPPAVTGITFLSGGQSEEEASINLNAINK
## [[&quot;P60709&quot;]] DLTDYLMKILTER
</code></pre>

<p>A Proteins instance is further described by general
<code>metadata</code>. Protein sequence and peptide features annotations can be
accessed with <code>ametadata</code> and <code>pmetadata</code> (or <code>acols</code> and <code>pcols</code>)
respectively.</p>

<pre><code class="r">metadata(p)
</code></pre>

<pre><code>## $created
## [1] &quot;Wed Jul 16 02:45:51 2014&quot;
</code></pre>

<pre><code class="r">head(acols(p))
</code></pre>

<pre><code>## DataFrame with 6 rows and 11 columns
##      DB AccessionNumber   EntryName IsoformName
##   &lt;Rle&gt;     &lt;character&gt; &lt;character&gt;       &lt;Rle&gt;
## 1    sp          A4UGR9 XIRP2_HUMAN          NA
## 2    sp          A6H8Y1  BDP1_HUMAN          NA
## 3    sp          O43707 ACTN4_HUMAN          NA
## 4    sp          O75369  FLNB_HUMAN          NA
## 5    sp          P00558  PGK1_HUMAN          NA
## 6    sp          P02545  LMNA_HUMAN          NA
##                                         ProteinName OrganismName GeneName
##                                         &lt;character&gt;        &lt;Rle&gt;    &lt;Rle&gt;
## 1     Xin actin-binding repeat-containing protein 2 Homo sapiens    XIRP2
## 2 Transcription factor TFIIIB component B&#39;&#39; homolog Homo sapiens     BDP1
## 3                                   Alpha-actinin-4 Homo sapiens    ACTN4
## 4                                         Filamin-B Homo sapiens     FLNB
## 5                         Phosphoglycerate kinase 1 Homo sapiens     PGK1
## 6                                      Prelamin-A/C Homo sapiens     LMNA
##            ProteinExistence SequenceVersion Comment
##                       &lt;Rle&gt;           &lt;Rle&gt;   &lt;Rle&gt;
## 1 Evidence at protein level               2      NA
## 2 Evidence at protein level               3      NA
## 3 Evidence at protein level               2      NA
## 4 Evidence at protein level               2      NA
## 5 Evidence at protein level               3      NA
## 6 Evidence at protein level               1      NA
##                                                                                        Filename
##                                                                                           &lt;Rle&gt;
## 1 /home/lgatto/R/x86_64-unknown-linux-gnu-library/3.2/Pbase/extdata/HUMAN_2014_06-09prots.fasta
## 2 /home/lgatto/R/x86_64-unknown-linux-gnu-library/3.2/Pbase/extdata/HUMAN_2014_06-09prots.fasta
## 3 /home/lgatto/R/x86_64-unknown-linux-gnu-library/3.2/Pbase/extdata/HUMAN_2014_06-09prots.fasta
## 4 /home/lgatto/R/x86_64-unknown-linux-gnu-library/3.2/Pbase/extdata/HUMAN_2014_06-09prots.fasta
## 5 /home/lgatto/R/x86_64-unknown-linux-gnu-library/3.2/Pbase/extdata/HUMAN_2014_06-09prots.fasta
## 6 /home/lgatto/R/x86_64-unknown-linux-gnu-library/3.2/Pbase/extdata/HUMAN_2014_06-09prots.fasta
</code></pre>

<pre><code class="r">head(pcols(p))
</code></pre>

<pre><code>## SplitDataFrameList of length 6
## $A4UGR9
## DataFrame with 36 rows and 47 columns
##        DB AccessionNumber   EntryName IsoformName
##     &lt;Rle&gt;     &lt;character&gt; &lt;character&gt;       &lt;Rle&gt;
## 1      sp          A4UGR9 XIRP2_HUMAN          NA
## 2      sp          A4UGR9 XIRP2_HUMAN          NA
## 3      sp          A4UGR9 XIRP2_HUMAN          NA
## 4      sp          A4UGR9 XIRP2_HUMAN          NA
## 5      sp          A4UGR9 XIRP2_HUMAN          NA
## ...   ...             ...         ...         ...
## 32     sp          A4UGR9 XIRP2_HUMAN          NA
## 33     sp          A4UGR9 XIRP2_HUMAN          NA
## 34     sp          A4UGR9 XIRP2_HUMAN          NA
## 35     sp          A4UGR9 XIRP2_HUMAN          NA
## 36     sp          A4UGR9 XIRP2_HUMAN          NA
##                                       ProteinName OrganismName GeneName
##                                       &lt;character&gt;        &lt;Rle&gt;    &lt;Rle&gt;
## 1   Xin actin-binding repeat-containing protein 2 Homo sapiens    XIRP2
## 2   Xin actin-binding repeat-containing protein 2 Homo sapiens    XIRP2
## 3   Xin actin-binding repeat-containing protein 2 Homo sapiens    XIRP2
## 4   Xin actin-binding repeat-containing protein 2 Homo sapiens    XIRP2
## 5   Xin actin-binding repeat-containing protein 2 Homo sapiens    XIRP2
## ...                                           ...          ...      ...
## 32  Xin actin-binding repeat-containing protein 2 Homo sapiens    XIRP2
## 33  Xin actin-binding repeat-containing protein 2 Homo sapiens    XIRP2
## 34  Xin actin-binding repeat-containing protein 2 Homo sapiens    XIRP2
## 35  Xin actin-binding repeat-containing protein 2 Homo sapiens    XIRP2
## 36  Xin actin-binding repeat-containing protein 2 Homo sapiens    XIRP2
##              ProteinExistence SequenceVersion Comment  spectrumid
##                         &lt;Rle&gt;           &lt;Rle&gt;   &lt;Rle&gt; &lt;character&gt;
## 1   Evidence at protein level               2      NA   index=124
## 2   Evidence at protein level               2      NA    index=28
## 3   Evidence at protein level               2      NA    index=20
## 4   Evidence at protein level               2      NA    index=21
## 5   Evidence at protein level               2      NA   index=187
## ...                       ...             ...     ...         ...
## 32  Evidence at protein level               2      NA    index=87
## 33  Evidence at protein level               2      NA    index=99
## 34  Evidence at protein level               2      NA     index=9
## 35  Evidence at protein level               2      NA   index=122
## 36  Evidence at protein level               2      NA    index=77
##     acquisitionnum passthreshold      rank calculatedmasstocharge
##          &lt;numeric&gt;     &lt;logical&gt; &lt;integer&gt;            &lt;character&gt;
## 1              124          TRUE         1        715.03076171875
## 2               28          TRUE         1      715.4117431640625
## 3               20          TRUE         1      786.9080810546875
## 4               21          TRUE         1       524.941162109375
## 5              187          TRUE         1      629.3385620117188
## ...            ...           ...       ...                    ...
## 32              87          TRUE         1      720.3527221679688
## 33              99          TRUE         1      618.7781982421875
## 34               9          TRUE         1          1013.51171875
## 35             122          TRUE         1       820.890869140625
## 36              77          TRUE         1      821.9254150390625
##     experimentalmasstocharge chargestate ms.gf.denovoscore ms.gf.evalue
##                  &lt;character&gt;   &lt;integer&gt;         &lt;numeric&gt;    &lt;numeric&gt;
## 1          715.0304565429688           3               146     0.002078
## 2          715.9176635742188           2                12     0.065243
## 3          786.9065551757812           2                35     0.119470
## 4          524.9410400390625           3                49     0.524066
## 5          629.8379516601562           2                53     0.339940
## ...                      ...         ...               ...          ...
## 32         720.3445434570312           2                63        42.38
## 33          619.288818359375           2                52        21.54
## 34        1014.0198364257812           2                17        21.91
## 35         821.4005126953125           2                56        22.61
## 36          821.923095703125           2               102        42.87
##     ms.gf.rawscore ms.gf.specevalue assumeddissociationmethod
##          &lt;numeric&gt;        &lt;numeric&gt;               &lt;character&gt;
## 1               36        1.823e-07                       CID
## 2              -62        5.738e-06                       CID
## 3              -44        1.050e-05                       CID
## 4              -19        4.606e-05                       CID
## 5              -33        2.993e-05                       CID
## ...            ...              ...                       ...
## 32             -90         0.003725                       CID
## 33             -63         0.001895                       CID
## 34            -148         0.001922                       CID
## 35             -97         0.001987                       CID
## 36            -102         0.003766                       CID
##     ctermioncurrentratio explainedioncurrentratio isotopeerror
##              &lt;character&gt;              &lt;character&gt;  &lt;character&gt;
## 1             0.18311569                0.3733551            0
## 2             0.17072089               0.17917182            1
## 3            0.081987105               0.15870962            0
## 4             0.05622067               0.14561509            0
## 5            0.082444645               0.16891155            1
## ...                  ...                      ...          ...
## 32           0.037362944              0.063179314            0
## 33           0.062403493               0.09917885            1
## 34           0.062099736               0.11363181            1
## 35           0.030872315              0.047606274            1
## 36           0.031126814              0.036955375            0
##     meanerrorall meanerrortop7 meanrelerrorall meanrelerrortop7
##      &lt;character&gt;   &lt;character&gt;     &lt;character&gt;      &lt;character&gt;
## 1      57.312866     55.738667       14.757639        46.320156
## 2      340.82858     151.35628        319.4936       126.973404
## 3      48.116673     48.116673       2.6420627        2.6420627
## 4      132.30872     132.30872      0.31386718       0.31386718
## 5      37.566624     37.566624        8.980245         8.980245
## ...          ...           ...             ...              ...
## 32     241.91089     241.91089       148.93636        148.93636
## 33     247.79701     247.79701      -103.02391       -103.02391
## 34     117.01554     117.01554       117.01554        117.01554
## 35     55.865704     55.865704        52.49654         52.49654
## 36     276.21368     276.21368       162.91066        162.91066
##     ms2ioncurrent ntermioncurrentratio nummatchedmainions stdeverrorall
##       &lt;character&gt;          &lt;character&gt;        &lt;character&gt;   &lt;character&gt;
## 1       1984657.5            0.1902394                 10      87.09684
## 2       1164420.1          0.008450932                  8      523.5172
## 3        439678.0            0.0767225                  7      34.59512
## 4       461579.66           0.08939442                  5      93.69792
## 5       1574020.8           0.08646689                  6     36.306866
## ...           ...                  ...                ...           ...
## 32       935006.0           0.02581637                  6     374.24667
## 33      1784755.5           0.03677536                  7     319.73682
## 34      317826.88           0.05153207                  5      76.52627
## 35      1221484.8          0.016733961                  3      68.27473
## 36      829413.75         0.0058285617                  4     213.22662
##     stdeverrortop7 stdevrelerrorall stdevrelerrortop7             pepseq
##        &lt;character&gt;      &lt;character&gt;       &lt;character&gt;        &lt;character&gt;
## 1         92.53631        103.21258          97.59206 QEITQNKSFFSSVKESQR
## 2         161.3231         536.8037         181.13979       LPVPKDVYSKQR
## 3         34.59512        59.203514         59.203514      EQNNDALEKSLRR
## 4         93.69792        162.12587         162.12587      EQNNDALEKSLRR
## 5        36.306866        51.466446         51.466446         SLKESSHRWK
## ...            ...              ...               ...                ...
## 32       374.24667         419.9993          419.9993      TNTSTGLKMAMER
## 33       319.73682        391.17908         391.17908       PESGFAEDSAAR
## 34        76.52627         76.52627          76.52627 QPDAIPGDIEKAIECLEK
## 35        68.27473         70.89802          70.89802     MARYQAAVSRGDCR
## 36       213.22662         308.5769          308.5769     QEITQNKSFFSSVK
##      modified      modification   isdecoy        post         pre
##     &lt;logical&gt;       &lt;character&gt; &lt;logical&gt; &lt;character&gt; &lt;character&gt;
## 1       FALSE                NA     FALSE           D           K
## 2       FALSE                NA     FALSE           N           R
## 3       FALSE                NA     FALSE           L           R
## 4       FALSE                NA     FALSE           L           R
## 5       FALSE                NA     FALSE           E           K
## ...       ...               ...       ...         ...         ...
## 32      FALSE                NA     FALSE           S           K
## 33      FALSE                NA     FALSE           G           K
## 34       TRUE 57.021463735 (15)     FALSE           A           K
## 35       TRUE 57.021463735 (13)     FALSE           S           R
## 36      FALSE                NA     FALSE           E           K
##           end     start    length                  spectrumFile
##     &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;                   &lt;character&gt;
## 1        2760      2743      3374 Thermo_Hela_PRTC_selected.mgf
## 2         318       307      3374 Thermo_Hela_PRTC_selected.mgf
## 3        1870      1858      3374 Thermo_Hela_PRTC_selected.mgf
## 4        1870      1858      3374 Thermo_Hela_PRTC_selected.mgf
## 5        1708      1699      3374 Thermo_Hela_PRTC_selected.mgf
## ...       ...       ...       ...                           ...
## 32       2094      2082      3374 Thermo_Hela_PRTC_selected.mgf
## 33         31        20      3374 Thermo_Hela_PRTC_selected.mgf
## 34       1729      1712      3374 Thermo_Hela_PRTC_selected.mgf
## 35         61        48      3374 Thermo_Hela_PRTC_selected.mgf
## 36       2756      2743      3374 Thermo_Hela_PRTC_selected.mgf
##                    databaseFile
##                     &lt;character&gt;
## 1   HUMAN_2014_06-09prots.fasta
## 2   HUMAN_2014_06-09prots.fasta
## 3   HUMAN_2014_06-09prots.fasta
## 4   HUMAN_2014_06-09prots.fasta
## 5   HUMAN_2014_06-09prots.fasta
## ...                         ...
## 32  HUMAN_2014_06-09prots.fasta
## 33  HUMAN_2014_06-09prots.fasta
## 34  HUMAN_2014_06-09prots.fasta
## 35  HUMAN_2014_06-09prots.fasta
## 36  HUMAN_2014_06-09prots.fasta
## 
## ...
## &lt;5 more elements&gt;
</code></pre>

<p>Specific proteins can be extracted by index of name using
<code>[</code> and proteins and their peptide features can be plotted
with the default plot method.</p>

<pre><code class="r">seqnames(p)
</code></pre>

<pre><code>## [1] &quot;A4UGR9&quot;   &quot;A6H8Y1&quot;   &quot;O43707&quot;   &quot;O75369&quot;   &quot;P00558&quot;   &quot;P02545&quot;  
## [7] &quot;P04075&quot;   &quot;P04075-2&quot; &quot;P60709&quot;
</code></pre>

<pre><code class="r">plot(p[c(1,9)])
</code></pre>

<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAfgAAAH4CAMAAACR9g9NAAACDVBMVEVEov95vP+JiYmSkpKUlJSZzP+cnJyczv+dnZ2fn5+goKChoaGioqKjo6OlpaWmpqapqamprq+psLKqqqqqrq+qr7Cqr7GqsLKqsbOqsrSqsrWqs7Wqs7aqs7eqtLeqtLiqtbiqtbmqtrmqtrqqt7uquL2qub6qusCqu8CqvMKqvsSqv8aq1f+rq6urt7yrub6rur6rvcOrvsSrvsWrv8arwcirw8qrw8urxMyrxc2rxs+rytOr1f+srKysxc2sxs+sx9CsyNKsydKsydOsytSsy9Wsy9aszNaszNeszdeszdisztmsz9qsz9us0Nus0Nys0+Cs1OGs1eKtra2t0d2t0t6t0+Ct1OCt2Oaurq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e32/+4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHB4P/CwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj4/P/5+fn5/P/6+vr7+/v8/Pz9/f3+/v7///+TD/uWAAAACXBIWXMAAAsSAAALEgHS3X78AAAOm0lEQVR4nO3djV8bdwHH8fpUp5srbj51D7Wpmw/Tuc3nTefD1Dqn86FM13aWarVPwJw2lG0tZQECpEChpWsgTZtQIEZI7vc3GhIu3GUpuTPQ4+77+byaex0XQn7kXQKX/vh1lyHJdgU9AAom4EUDXrRN4S/sLpvT6cQ7xnTN36sB0VZlnXpjzz+s8smjfzm6UNs6JTeHP3Ya+NA2WHG7kIsPGDP3dG3rHT55erwCf19X10eBD10n5ta2Ty9UNp+sbZ2SLeDLezrTiX6+4sPYQMWtd6r6td5Z2zolW8Cb2x8GPqRZJzpf/JtVPnX0jSNLta1neIpuwIsGvGjAiwa8aCGAt9J7TO7hIwP2JejxBN/1jq7+dh+REMAv9t1vzk6a3fYl6PEE37t564F2H5EQwBvzoDmaMw8+vH5ZDXo4wZf9RG+7j0hI4M9Mmd32JejRBN/QirnvdJuPSEjgcy92XbQvQY8m+FIdJ3rbfURCAU9bH/CihQd+8YYV9BB2WDezbdw4PPDJjlLQQ9hhne1t48bAh7dths8M7oy6Oy4FPYQd1rHj/m9T8A6f6qDolANeMx/w2SS5On0k6BG0UdE7PDX0dlfQI9iKgPcd8KIt3g56BFsR8KIBLxrw3pvvOH4mMnPAgPfepYz51MmozAED3kfFf52LzBww4L2XOZ6Jzhww4L331z2dnTejMgcMeNGAFw140YAXDXjRgBcNeNGAF20d/uZM8+4EOzrattbhryWadyvY0dG2BbxowIsGvGjAiwa8aMCLBrxowIsGvGit4T+0i6LTN7zDUyQDXjTgRQNeNOBFA1404EUDXjTgRQNetHX44mLzWDc4qq3Dpy83Lx/s6Gjb4qleNOBFA1404EUDXjTgRQNeNOBFA1404EUDXjTgRQNeNOBFA1404EUDXjTgRQNeNOBFA140L/CFeLyQjcfzZib+rumPj5jJeL/V5GNdjN/wcc+Z+MxkfLC6+178mvu6K/FJf5/GaHzJrH+o+PvVnavx+dqRm/G1j36x+ZAdzcX74/GrjUcn4u/Vdgbiy/WDg/Xh9cevexlef8vPZ7UvPmzvl8/H48NX433rc9uL8XjSy534ywt89rnnslM//u6sufjKvnLsULc5/7tYucnH2v+SnwFOPHWh77Xu6m73wUH3dW//8Ly/T6PnW5naTvehn9cewIGnUuv38/yvLlXGdqjpkB2N/+DYay9/4L8NPfen2hDN/q/drB/sfq2vtlOO/SLhYXTl2O/7WrzL4rOHuu39whOHDnUPvPzk+t/l3Fc3rtq6PME/9VR26iffq8BXwGOHe8z5P3+p2aMY+6Uv+K9f6Hu9p7rb80oj/K/9wn9nHb7n8MF1+K/b8D/9TQU+drjpkB2Nv3Ds9Vc/CP/H2hBN7Nsb8D2v1+Ff9gb/h5bwXzncY+8Xnjx8uGfg1S/b8M9tXLV1AW8HPPCOgAceeOCBdwX8XQLeGfDOgK8GPPDeAt5vwNsBD7wj4IEHHnjgXQF/l4B3Brwz4KsBD7y3gPcb8HbAA+8IeOCBBx54V8DfJeCdAe8M+GrAA+8t4P0GvB3wwDsCHnjggQfeFfB3CXhnwDsDvhrwwHsLeL8Bbwc88I6ABx544IF3BfxdAt4Z8M6ArwY88N4C3m/A2wEPvCPggdeDzz/6aH5232NzZmT/49beA31mMLa32VLQ+/dN+rjnq48MX4q9Wd19a/+Y+7rEvsEmt9ik84/cru28dWDfRHUn+ch07ci1x/ePGvPFA02H7OjKvpOx/R/4mzsQe6u2E3s0Xz/4ZuxSbcfauy/lYXTWXvsGd62w98A5e39l74ED55L79xZrby6uPeRbHuvViwa8aMCLBrxowIsGvGjAiwa8aMCLBrxowIsGvGjAiwa8aMCLBrxowIsGvGjAiwa8aMCLBrxowIsGvGjAiwa8aMCLBrxowIsGvGjAiwa8aMCLBrxowIsGvGjAiwa8aA3ww2OJ8dkU8NHPDZ9azY1ZxpoGPvK54fMmkzEL1hLwkc8NX7qTKJSH8mXgI58bvpwbNblEvgR85HPDL5qiuTrDU71AbvjJslkZypqZ0MGXVtr+EFax2fLW5WKptrPa9B5Wiy2WRq5/8FbvsWLfUf2O6+NpvGpLajidS06NJKYnw3c61/vQzdbvtHmXH4s3Odr3xKnazhceavJYWJ+ODXv64MOfvbb5Oyx97pmzrgOXYt/sr+0tfP6Z7V+2fHgmPZ4aDh98fb36/7/xF/7d5GjT9erreVyv3pjEszObv4NzvfpqA6/aa9zfi/XqRyvPSFPzxVHg7UTgb5s7ZmrW3ALeTgS+tDBiphL5VeDtVODza/C58J3HA++3hpdsrazJZM0C8HYi8GOra2+UxoG3E4FPDF/P3roxwulcPRX4xoAXgC9sBLydArzjKPB2CvCpVOqayU5OZs0N4O0U4CstL1e3PNXXE4EvrQxXfrJf4QWceiLwOVNIpwsmD7ydCHxyee0N/nVuIxH4xNB0em56iPP4eirwvIDTkAK8tZyw1gPeTgHeFOqfBPB2CvBDQ5U/tYC3U4CvNDO2tr18BXg7CfiJCZOeqLQQvl+hAt5vTvj6wRXg7STgl5ZMcanSHZ7q60nAV+g3yIGvJgKfSAynnL9IA7wI/FDGMsaa53Sungj8vLGWC8bMAW8nAr+yWjmRv1zmp/p6KvDV35pbLAJvJwL/fmkikbhizQJvJwJfNKZUMlaxEvDVROAdVwBfTQQ+tRHw1UTgWQqlMRF4lkJpTASepVAaE4FnKZTGVOBZCqUhEXiWQmlMBJ6lUBoTgWcplMZU4JmI0ZAIfGgnYnQ/2z78t5vCv3SmtvPYM83gH/m+R/hYK/jYQfcixgPP/8yGf+Jgt6c78VVEJmJcG2u5LnircskbTY6+n5yu7UyONVu3fDzp7RHKji5u/g6l0eRV14FMMpmu7a2OJlssff3/FJGJGOS3iEzEIL9FZCIG+S0iEzHIbxGZiEF+i8hEDPJbRJZCIb8BLxrwogEvGvCiAS8a8KIBLxrwogEvGvCiAS8a8KIBLxrwoinD53JBjyDAlOHz+aBHEGDK8NIBLxrwogEvGvCiAS8a8KIBLxrwogEvGvCiAS8a8KIBLxrwogEvGvCiAS8a8KIBLxrwogG/MyolBqfu6R0CvzNa/OZvz7Z+ry0M+J3RB/6jgu0O+J0R8KIBLxrwogEvGvCiAS8a8KIBLxrwogEvGvCiAS8a8KIBLxrwogEvGvCiAS8a8KIBLxrwogEvGvCiAS8a8KIBLxrwogEvGvCiAS8a8KIBLxrwogEvGvCiAS8a8KIBLxrwogEvGvCiAS8a8KItfebAmXt6h8CLBrxowIsGvGjAiwa8aMCLBrxowIsGvGjAiwa8aMCLBrxowIsGvGjAiwa8aMCLBrxom8If+8hNczqdeMeYrvlgh0n+s069secfVvnk0b8cXahtnZKbw99/HPjQNlhxu5CLDxgz93Rt2wTeKjfr3ZHT4xX4+7q6Pgp86Doxt7Z9eqGy+WRt65TctdlNLyTLezrTiX6+4sPYQMWtd6r6td5Z2zolW8Cb2x8GPqRZJzpf/JtVPnX0jSNLta1neIpuwIsGvGjAiwa8aMCLFgJ4K73H5B4+MmBfgh5P8F3v6Opv9xEJAfxi3/3m7KTZbV+CHk/wvZu3Hmj3EQkBvDEPmqM58+DD65fVoIcTfNlP9Lb7iIQE/syU2W1fgh5N8A2tmPtOt/mIhAQ+92LXRfsS9GiCL9VxorfdRyQU8LT1AS9aeOBnestBD2GHdXGwjRuHBz7ZUQp6CDuss71t3Bj48LbN8KkOik454DUDXjQf8NTQ211Bj2ArAt53wIs2/FbQI9iKgBcNeNGA9958x/EzkZkKBLz3LmXMp05GZSoQ8D4q/utcZKYCAe+9zPFMdKYCAe+9v+7p7LwZlalAwIsGvGjAiwa8aMCLBrxowIsGvGjAi8ZatqIBLxrwogEvGvCiAS8a8KIBLxrwogEvWmv4v3/Mdx//T7CfFLWuNfw/v+G7H/032E+KWsdTvWjAiwa8aMCLBrxowIsGvGjAiwa8aMCLtg5fXGpeqBd9oE3iK1404EUDXjTgRQNeNOBFA1404EUDXjTgRQNeNOBFA1404EUDXrQG+OGxxPhsCvjo54ZPrebGLGNNAx/53PB5k8mYBWsJ+Mjnhi/dSRTKQ/ky8JHPDV/OjZpcIl8CPvK54RdN0Vyd4aleIDf8ZNmsDGXNjCf4UrF0T4ZI21HD6VxyaiQxPentdK73mZ57MkTajhrP42fS46lhb/A9h4EPb2740aIxU/PFUeAjnxv+trljpmbNLeAjX8N5/MKImUrkV4GPfA3w+TX4nLfzeODDXMNLtlbWZLJmAfjI54Yfq/52bGkc+MjXeDp3PXvrxginc9GvjYkYwIc5B3xhI+AjnwPecRT4yOeAT6VS10x2cjJrbgAf+dzf45eXq1ue6qNfwws4K8OVn+xXeAEn+rnhc6aQThdMHvjI54ZPLq+9wb/OCdRwHj80nZ6bHuI8PvrxAo5oDnhrOWGtB3zkc76AU6i/hgN85HPADw1V/tQCPvK5v8fPjK1tL18BPvI54ScmTHqi0oK3X6ECPsw54esHV4CPfE74paXauvV3eKqPfu7v8Usb5MBHu8YXcIZTzl+kAT6yueGHMpYx1jync9HPDT9vrOWCMXPARz43/Mpq5UT+cpmf6qNfA3z1t+YWi8BHPjf8+6WJROKKNQt85HPDF40plYxVrAR8tHPDO64APto1LHC4EfDRjqVQRGMpFNFYCkU0lkIRjaVQRGMpFNFYCkU0lkIRjYkYorUxEaP7YPc9GSJtR21MxLiWnL0nQ6TtqI2JGBTm2piIQWGujYkYFObamIhBYa6NiRgU5tqYiEFhjv9UWDTgRQNeNOBFA1404EUDXjTgRQNeNOBFA1404EUDXjTgRQNetHX47GzzFoMdHW1bu4IeAAUT8KIBLxrwogEvGvCi/Q+r8pOxmXvE2AAAAABJRU5ErkJggg==" title="plot of chunk pplot" alt="plot of chunk pplot" style="display: block; margin: auto;" /></p>

<p>More details can be found in <code>?Proteins</code>. The object generated above
is also directly available as <code>data(p)</code>.</p>

<h2>Session information</h2>

<pre><code class="r">sessionInfo()
</code></pre>

<pre><code>## R Under development (unstable) (2014-04-10 r65396)
## Platform: x86_64-unknown-linux-gnu (64-bit)
## 
## locale:
##  [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
##  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
##  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] grid      parallel  methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
## [1] Pbase_0.1.7           Gviz_1.9.10           GenomicRanges_1.17.23
## [4] GenomeInfoDb_1.1.12   IRanges_1.99.23       S4Vectors_0.1.2      
## [7] Rcpp_0.11.2           BiocGenerics_0.11.3   knitr_1.6            
## 
## loaded via a namespace (and not attached):
##  [1] affy_1.43.3               affyio_1.33.0            
##  [3] AnnotationDbi_1.27.8      BatchJobs_1.3            
##  [5] BBmisc_1.7                Biobase_2.25.0           
##  [7] BiocInstaller_1.15.5      BiocParallel_0.7.8       
##  [9] biomaRt_2.21.1            Biostrings_2.33.12       
## [11] biovizBase_1.13.8         bitops_1.0-6             
## [13] brew_1.0-6                BSgenome_1.33.8          
## [15] checkmate_1.2             cleaver_1.3.7            
## [17] cluster_1.15.2            codetools_0.2-8          
## [19] colorspace_1.2-4          data.table_1.9.2         
## [21] DBI_0.2-7                 dichromat_2.0-0          
## [23] digest_0.6.4              doParallel_1.0.8         
## [25] evaluate_0.5.5            fail_1.2                 
## [27] foreach_1.4.2             formatR_0.10             
## [29] Formula_1.1-2             GenomicAlignments_1.1.20 
## [31] GenomicFeatures_1.17.12   ggplot2_1.0.0            
## [33] gtable_0.1.2              Hmisc_3.14-4             
## [35] impute_1.39.0             iterators_1.0.7          
## [37] lattice_0.20-29           latticeExtra_0.6-26      
## [39] limma_3.21.10             MALDIquant_1.10          
## [41] MASS_7.3-33               matrixStats_0.10.0       
## [43] MSnbase_1.13.12           munsell_0.4.2            
## [45] mzID_1.3.4                mzR_1.11.10              
## [47] pcaMethods_1.55.0         plyr_1.8.1               
## [49] preprocessCore_1.27.1     proto_0.3-10             
## [51] Pviz_0.99.0               R.methodsS3_1.6.1        
## [53] RColorBrewer_1.0-5        RCurl_1.95-4.1           
## [55] reshape2_1.4              Rsamtools_1.17.31        
## [57] RSQLite_0.11.4            rtracklayer_1.25.13      
## [59] scales_0.2.4              sendmailR_1.1-2          
## [61] splines_3.2.0             stats4_3.2.0             
## [63] stringr_0.6.2             survival_2.37-7          
## [65] tools_3.2.0               VariantAnnotation_1.11.16
## [67] vsn_3.33.0                XML_3.98-1.1             
## [69] XVector_0.5.7             zlibbioc_1.11.1
</code></pre>

</body>

</html>
